Add get_latest_data_path() and download_latest_data()

LiamHz opened the pull request:
In the S3 public bucket, we're adding new lightweight data files for:
- Test
- Live
- Validation
- Latest test era
- Tournament IDs

These are present for the most recent week, and are marked as "latest". E.g. `latest_numerai_live_data.csv`

This is so users don't have to download the entire dataset every time. ETA for these changes to go live is Monday.

LiamHz committed:
Add function and tests for get_latest_data_path()

Also create convenience functions for get_latest_data_path() such as
get_latest_live_data_path()

uuazed commented on numerapi/numerapi.py:
```diff
@@ -112,6 +114,73 @@ def download_current_dataset(self, dest_path=".", dest_filename=None,
+
+        Args:
+            data_type (str): type of data to return
+            data_type (str): file extension to get (optional, defaults to csv)
```
data_type arg twice

uuazed commented on numerapi/numerapi.py:
```diff
@@ -112,6 +114,73 @@ def download_current_dataset(self, dest_path=".", dest_filename=None,
+        ]
+
+        # Allow ext to have a "." as the first character
+        if ext[0] == ".":
```
maybe just do `ext.lstrip(".")`

uuazed commented on numerapi/numerapi.py:
```diff
@@ -112,6 +114,73 @@ def download_current_dataset(self, dest_path=".", dest_filename=None,
 
         return dataset_path
 
+    def get_latest_data_path(self, data_type: str, ext: str = "csv") -> str:
```
I prefer using full words not abbreviations. Please rename to `extension`

uuazed replied:
maybe rename from `path` to `url`, to make it more explicit that this needs to be downloaded

uuazed commented on numerapi/numerapi.py:
```diff
@@ -112,6 +114,73 @@ def download_current_dataset(self, dest_path=".", dest_filename=None,
+            str: url of the requested dataset
+
+        Example:
+            >>> NumerAPI().get_latest_data_path("live", "csv")
```
in `utils` there is function to download files from urls. Maybe add that to the example like `numerapi.utils.dowload_file(url)`

uuazed commented on tests/test_numerapi.py:
```diff
@@ -8,6 +8,8 @@
 import numerapi
 from numerapi import base_api
 
+PUBLIC_DATASETS_URL = "https://numerai-public-datasets.s3-us-west-2.amazonaws.com"
```
re-use the definition from numerapi.py.

uuazed commented on tests/test_numerapi.py:
```diff
@@ -36,6 +38,36 @@ def test_download_current_dataset(api, tmpdir):
         assert len(rsps.calls) == 0
 
 
+def test_get_latest_data_paths(api):
```
you could convert that to a pytest parametrized test, like this one: https://github.com/uuazed/numerapi/blob/9690012a8753f82373703e7041959c0d7adac31e/tests/test_numerapi.py#L49. But it's also fine like this.

uuazed reviewed (changes requested):
Thx for your contribution, looking forward to start using it. 

I left a few comments, but nothing major. Please address those and I'll merge your change

LiamHz committed:
Rename current_test_era to max_test_era

LiamHz committed:
Implement requested changes from uuazed

- Rename ext to extension
- Rename get_latest_data_path to get_latest_data_url
- Minor: using lstrip(), adding download example, typo fix

LiamHz committed:
Make convenince functions download latest data

LiamHz commented on numerapi/numerapi.py:
```diff
@@ -155,30 +155,35 @@ def get_latest_data_url(self, data_type: str, extension: str = "csv") -> str:
-    def get_latest_training_data_url(self, extension: str = "csv") -> str:
-        return self.get_latest_data_url("training", extension)
+    # Convenience functions for download_latest_data()
+    # The extension argument must be one of: "csv", "csv.xz", or "parquet"
```
It might make sense to delete these "convenience" functions and only keep the new function `download_latest_data()`

uuazed replied:
I agree

LiamHz replied:
Done!

LiamHz commented on tests/test_numerapi.py:
```diff
@@ -36,33 +36,33 @@ def test_download_current_dataset(api, tmpdir):
-        "example_predictions": api.get_latest_example_predictions_data_url,
-    }
+def test_get_latest_data_url(api):
+    # List of data types that have latest data files
```
Should we add a new test similar to `test_download_current_dataset()`, but for `download_latest_data()`?

uuazed replied:
I'll merge the PR without that test ;-)

LiamHz committed:
Remove convenience functions
